pyk/src/pyk/k2lean4/Prelude.lean

```diff
@@ -1,10 +1,105 @@
-abbrev SortBool : Type := Int
-abbrev SortBytes: Type := ByteArray
-abbrev SortId : Type := String
-abbrev SortInt : Type := Int
-abbrev SortString : Type := String
+/-
+K Prelude in Lean 4
+
+Functions with the `hook` attribute need to have a manual implementation in the backends.
+This file contains the Lean 4 definitions of the hooked functions in `domains.md`.
+
+Currently we translate hooked functions as uninterpreted functions together with axioms asserting their behavior.
+The current definition can be put into three levels:
+
+1. Signature Level:
+The signature of the hooks, this includes aliases for Sorts and function symbols for hooked functions.
+
+2. Rule Level:
+The behavior of the uninterpreted symbols can be asserted through axioms or theorems.
+Inconsistencies can arise from them, so it falls under the user to make sure axioms are consistent and/or theorems provable.
+
+3. Simplification Level:
+With the theory defined through function rules, simplifications can be stated as theorems.
+These theorems should be provable directly from the function rules and the semantics of the Sorts.
+ -/
+
+-- Basic K types
+abbrev SortBool         : Type := Int
+abbrev SortBytes        : Type := ByteArray
+abbrev SortId           : Type := String
+abbrev SortInt          : Type := Int
+abbrev SortString       : Type := String
 abbrev SortStringBuffer : Type := String
 
 abbrev ListHook (E : Type) : Type := List E
-abbrev MapHook (K : Type) (V : Type) : Type := List (K × V)
 abbrev SetHook (E : Type) : Type := List E
+
+namespace MapHookDef
+/-
+The `Map` sort represents a generalized associative array.
+Each key can be paired with an arbitrary value, and can be used to reference its associated value.
+Multiple bindings for the same key are not allowed.
+Note that both keys and values will always be KItems.
+ -/
+
+-- Signature to be instantiated by map implementations
+structure MapHookSig (K V : Type) where
+  map        : Type -- Carrier, such as List (KItem × KItem)
+  unit       : map
+  cons       : K → V → map → map
+  lookup     : map → K → V
```

Comment:
Is this not a non-total function?

```suggestion
  lookup     : map → K → Option V
```

(The `Ceil` rules in `domains.md` should be reviewed for all these functions to see which ones are non-total.)

Reply:
As per [this description](https://github.com/runtimeverification/k/blob/master/k-distribution/include/kframework/builtin/domains.md#map-lookup) I thought it should be implemented as a `lookup!`, which I believe has the current type. But if that's not what we want we can have it with the `Option` type

Reply:
I'm not sure what `lookup!` is, is that a LEAN4 thing? I guess initially I would agree that anything non-total should go to `Option`, but probably `Option` is also not quite what we want. It would be best if we could represent `#Bottom` directly, maybe the discussion with the Pi2 folks who implemented ML in LEAN4 will help clarify things.

Reply:
The notation for `!` after a function indicates that it will throw a panic error if the argument it is called with is not defined, but I guess this is not the same as returning `Bottom`.

Reply:
Perhaps we can have it as a sum type? Something like
```lean
axiom lookupAx : mapCAx → K → V ⊕ Empty
axiom lookup_def : forall m k, in_keysAx K m k → ∃ v, lookupAx K V m k = v
```
This way, if it's defined, we know it'll have the correct value, and otherwise we get a contradiction

Reply:
It can just return an `Option` for now.